lib/storage/storage.go

```diff
@@ -2819,6 +2827,81 @@ func (s *Storage) mustOpenIndexDBTables(path string) (next, curr, prev *indexDB)
 	return next, curr, prev
 }
 
+func (s *Storage) startReadonlyWatchdog() {
+	s.readonlyWatchdogWG.Add(1)
+	go func() {
+		s.readonlyWatchDog()
+		s.readonlyWatchdogWG.Done()
+	}()
+}
+
+func (s *Storage) readonlyWatchDog() {
+	timeout := 10 * time.Second
+	for {
+		if !s.cannotWrite.Load() {
+			timeoutChan := time.After(timeout)
+			tmpFileChan := make(chan *os.File, 1)
+			go func() {
+				tmpFile, err := os.CreateTemp(s.path, ".is_readonly_watchdog")
+				if err != nil {
+					tmpFileChan <- nil
+				} else {
+					tmpFileChan <- tmpFile
+				}
+			}()
+
+			select {
+			case tmpFile := <-tmpFileChan:
+				if tmpFile == nil {
+					logger.Panicf("cannot create files on the disk")
+				} else {
+					if !MustBeWritable(tmpFile) {
+						logger.Panicf("cannot write files to the disk")
+					}
+				}
+			case <-timeoutChan:
+				logger.Panicf("cannot write to the disk or disk is overloaded and takes more than %s to write", timeout.String())
+			}
+
+			time.Sleep(5 * time.Second)
+		}
+	}
+}
+
+func MustBeWritable(tmpFile *os.File) bool {
+	defer tmpFile.Close()
+
+	totalSize := 10 * 1024 * 1024 // 10 MB
```

Comment:
Also I'm not sure we need to write every 5 seconds. Might be a bit too much, no?

Reply:
Yes, it feels like this implementation is a bit flawed, though it does solve the problem.